tests/remote_test.go

```diff
@@ -0,0 +1,143 @@
+// limitations under the License.
+
+package tests
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	ptesting "github.com/pulumi/pulumi/sdk/v3/go/common/testing"
+)
+
+const remoteTestRepo = "https://github.com/pulumi/test-repo.git"
+
+func TestInvalidRemoteFlags(t *testing.T) {
+	t.Parallel()
+
+	commands := []string{"preview", "up", "refresh", "destroy"}
+
+	tests := map[string]struct {
+		args []string
+		err  string
+	}{
+		"no url": {
+			err: "error: must specify remote URL",
+		},
+		"no branch or commit": {
+			args: []string{remoteTestRepo},
+			err:  "error: either `--remote-git-branch` or `--remote-git-commit` is required",
+		},
+		"both branch and commit": {
+			args: []string{remoteTestRepo, "--remote-git-branch", "branch", "--remote-git-commit", "commit"},
+			err:  "error: `--remote-git-branch` and `--remote-git-commit` cannot both be specified",
+		},
+		"both ssh private key and path": {
+			args: []string{remoteTestRepo, "--remote-git-branch", "branch", "--remote-git-auth-ssh-private-key", "key",
+				"--remote-git-auth-ssh-private-key-path", "path"},
+			err: "error: `--remote-git-auth-ssh-private-key` and `--remote-git-auth-ssh-private-key-path` " +
+				"cannot both be specified",
+		},
+		"ssh private key path doesn't exist": {
+			args: []string{remoteTestRepo, "--remote-git-branch", "branch", "--remote-git-auth-ssh-private-key-path",
+				"doesntexist"},
+			err: "error: reading SSH private key path",
+		},
+		"invalid env": {
+			args: []string{remoteTestRepo, "--remote-git-branch", "branch", "--remote-env", "invalid"},
+			err:  `expected value of the form "NAME=value": missing "=" in "invalid"`,
+		},
+		"empty env name": {
+			args: []string{remoteTestRepo, "--remote-git-branch", "branch", "--remote-env", "=value"},
+			err:  `error: expected non-empty environment name in "=value"`,
+		},
+		"invalid secret env": {
+			args: []string{remoteTestRepo, "--remote-git-branch", "branch", "--remote-env-secret", "blah"},
+			err:  `expected value of the form "NAME=value": missing "=" in "blah"`,
+		},
+		"empty secret env name": {
+			args: []string{remoteTestRepo, "--remote-git-branch", "branch", "--remote-env-secret", "=value"},
+			err:  `error: expected non-empty environment name in "=value"`,
+		},
+	}
+
+	for _, command := range commands {
+		command := command
+		for name, tc := range tests {
+			tc := tc
+			t.Run(command+"_"+name, func(t *testing.T) {
+				t.Parallel()
+
+				e := ptesting.NewEnvironment(t)
+				defer deleteIfNotFailed(e)
+
+				// Remote flags currently require PULUMI_EXPERIMENTAL.
+				e.Env = append(e.Env, "PULUMI_EXPERIMENTAL=true")
+
+				args := []string{command, "--remote"}
+				_, err := e.RunCommandExpectError("pulumi", append(args, tc.args...)...)
+				assert.NotEmpty(t, tc.err)
+				assert.Contains(t, err, tc.err)
+			})
+		}
+	}
+}
+
+func TestRemoteLifecycle(t *testing.T) {
```

Comment:
Unfortunately, this test task ~3 minutes. I'm not sure if this will currently run as part of a PR or only in master, but ideally, we wouldn't hold up a PR over it. @AaronFriel, what do I need to do to make this not run in PRs?

Reply:
We'll disable the test by default (requires `PULUMI_TEST_DEPLOYMENTS_API` to be set). And consider running the test in a cron until we add support for mocking the deployments api.